src/server/frontend_wayland/text_input_v1.cpp

```diff
@@ -0,0 +1,421 @@
+{
+}
+
+TextInputV1::~TextInputV1()
+{
+    if (seat)
+    {
+        seat.value()->remove_focus_listener(client, this);
+        delete seat.value();
+    }
+
+    on_new_input_field = false;
+    pending_state.reset();
+}
+
+void TextInputV1::send_text_change(ms::TextInputChange const& change)
+{
+    auto const client_serial = find_client_serial(change.serial);
+    if (!pending_state || !current_surface || !client_serial)
+    {
+        // We are no longer enabled, or we don't have a valid serial
+        return;
+    }
+    if (change.preedit_text || change.preedit_cursor_begin || change.preedit_cursor_end)
+    {
+        send_preedit_cursor_event(change.preedit_cursor_begin.value_or(0));
+        send_preedit_string_event(client_serial.value(), change.preedit_text.value_or(""), "");
+    }
+    if (change.delete_before || change.delete_after)
+    {
+        send_delete_surrounding_text_event(change.delete_before.value_or(0), change.delete_after.value_or(0));
+    }
+    if (change.commit_text)
+    {
+        send_commit_string_event(client_serial.value(), change.commit_text.value());
+    }
+}
+
+auto TextInputV1::find_client_serial(ms::TextInputStateSerial state_serial) const -> std::optional<uint32_t>
+{
+    // Loop in reverse order because the serial we're looking for will generally be at the end
+    for (auto it = state_serials.rbegin(); it != state_serials.rend(); it++)
+    {
+        if (it->second == state_serial)
+        {
+            return it->first;
+        }
+    }
+    return std::nullopt;
+}
+
+void TextInputV1::focus_on(mf::WlSurface* surface)
+{
+    if (current_surface)
+    {
+        send_leave_event();
+    }
+    current_surface = mw::make_weak(surface);
+    if (surface)
+    {
+        send_enter_event(surface->resource);
+    }
+    else
+    {
+        deactivate(nullptr); // TODO - Is this right?
+    }
+}
+
+void TextInputV1::activate(wl_resource *seat, wl_resource *surface)
+{
+    (void)surface;
+
+    auto const wl_seat = mf::WlSeat::from(seat);
+    if (!wl_seat)
+    {
+        BOOST_THROW_EXCEPTION(std::runtime_error("failed to resolve WlSeat activating TextInputV1"));
+    }
+    this->seat = std::optional<mf::WlSeat*>(wl_seat);
+    this->seat.value()->add_focus_listener(client, this);
+
+    if (current_surface)
+    {
+        on_new_input_field = true;
+        pending_state.emplace();
+    }
+}
+
+void TextInputV1::deactivate(wl_resource *seat)
+{
+    (void)seat;
+    on_new_input_field = false;
+    pending_state.reset();
+}
+
+void TextInputV1::show_input_panel()
+{
+    commit_count++;
+    pending_state.emplace(); // TODO - This makes sure that pending_state is ALWAYS true, making the next check redundant. Consult with Sophie.
+
+    if (pending_state && current_surface)
+    {
+        auto const new_serial = ctx->text_input_hub->set_handler_state(handler, on_new_input_field, *pending_state);
+        state_serials.push_back({commit_count, new_serial});
```

Comment:
It appears this is incorrect, as is the code you copied it from. v3 uses the commit count as a serial, but both v2 and v1 use an explicit serial given by the client. This might mean you need to move this logic to `commit_state()`? Not sure.

Reply:
v2 fixed in #2444. You might be able to move all the logic here to `commit_state()` and leave `show_input_panel()` and `hide_input_panel()` empty (as in v2).

Reply:
It seems that `commit_state()` never gets called by Electron. Additionally, `commit_state()` seems to be misleadingly named.

This is the behavior of `commit_state()` in v1 according to the protocol:
`used to identify the known state`

This is opposed to `update_state()` in v2:
```
Allows to atomically send state updates from client.

This request should follow after a batch of state updating requests
like set_surrounding_text, set_content_type, set_cursor_rectangle and
set_preferred_language.
...
```
It seems their behavior is different. Any advice on how to proceed?

Reply:
> This is the behavior of `commit_state()` in v1 according to the protocol:
> `used to identify the known state`

"New serials are sent by the text input in the commit_state request and are used by the input method to indicate the known text input state in events like preedit_string, commit_string, and keysym. The text input can then ignore events from the input method which are based on an outdated state (for example after a reset)."

So it sounds like it is yet another way to manage the serial.